docs/todo-due-dates.md

```diff
@@ -0,0 +1,134 @@
+- Authors can specify the number of days the todos must be fixed after its creation date.
+- Authors can specify the number of days the todos are flagged as `error`s after they are past-due. The default value is `0`.
+- Todos that are past-due are flagged as `warning`s until the number of days to error is reached.
+- Configuration can be specified in a common file or overriden via CLI arguments.
+
+## Implementation
+
+### High Level Flow
+
+When an author runs a linter with the todo functionality, all errors are converted to todos with no due dates unless one is present in `package.json` or passed as a command line argument.
+
+- `daysToFix` is the number of days that the todo item needs to be fixed.
+- `daysToError` is the number of additional days past the `daysToFix` that the todo items are cosidered errors. Default is `0`.
+
+### Workflows
+
+#### Normal todo workflow
+
+- No due date is expected when adding todo items with no config.
+- No due date is expected when adding todo items with no CLI param.
+
+#### Todos that are past-due become errors
+
+Use case: 20 days for todos to be considered errors since its creation date.
+
+**Via config:**
+
+In `package.json`:
+
+ ```json
+{
+  "lint-todo": {
+    "daysToFix": 20,
+  }
+}
+```
+
+**Via `eslint` CLI:**
+
+```
+UPDATE_TODO=1 DAYS_TO_FIX_TODO=20 yarn eslint . --format eslint-formatter-todo
+```
+
+**Via `ember-template-lint` CLI:**
+
+```
+yarn ember-template-lint . --update-todo --days-to-fix-todo 20
+```
+
+#### Todos that are past-due become warnings, then errors
+
+Use case: give 20 days for todos to be considered `warning`s, then 5 additional days until they are considered `error`s.
+
+**Via config:**
+
+```json
+{
+  "lint-todo": {
+    "daysToFix": 20,
+    "daysToError": 5
+  }
+}
+```
+
+**Via `eslint` CLI:**
+
+```
+UPDATE_TODO=1 DAYS_TO_FIX_TODO=20 DAYS_TO_ERROR_TODO=5 yarn eslint . --format eslint-formatter-todo
+```
+
+**Via `ember-template-lint` CLI:**
+
+```
+yarn ember-template-lint . --update-todo --days-to-fix-todo 20 --days-to-error-todo 5
+```
+
+### Proposed changes to the schema to include due date
+
+```ts
+interface TodoData {
+  ...
+  created: Date;
+  daysToFix?: int;
+  daysToError?: int;
+}
+```
+
+# Critique
+
+[critique]: #critique
+
+A few other options were considered:
+
+- Instead of due days, another option considered was passing a specific due date. Passing days is preferable because if we want to expose a config/CLI param for this, passing a date would make the config or command to run stale.
+- Allow the option to pass rule id with due date associated to them. The extra rule id config is not necessary since both `ember-template-lint` and `eslint` already have options to filter by rule id.
+- If devs needs extra functionality, we could provide a different npx tool to do that. For now, the goal is to keep the todo functionality simple with a simple due date mechanism.
+
+# Unresolved questions
+
+[unresolved]: #unresolved-questions
```

Comment:
How will we ensure that if todos are completely regenerated (we run `UPDATE_TODO=1 or `--update-todo`) we don't lose this data?

Reply:
I guess I'm not 100% sure if we should update the dates. If an initial due date is set, I think that should stand going forward. What do you think?

So, I think the second would be better. What circumstance would you want to refresh the todo due date?

Reply:
I can see the situation of people wanting to push the due date further because they didn't have enough time to fix the errors within X days, so they want to refresh the dates of all todos. WDYT?

---

Here's the proposal for the implementation:

When calling `writeTodos()` authors can now optionally pass `daysToError` and `daysToWarn`:

```ts
export declare function writeTodos(baseDir: string, lintResults: LintResult[], filePath?: string, daysToError?: int, daysToWarn?: int): Promise<string>;
```

**writeTodos() expected implementation:**

Writes files for todo lint violations. One file is generated for each violation, using a generated hash to identify each.

Given a list of todo lint violations, this function will also:

  - Delete existing files that no longer have a todo lint violation.
  - Keep existing files untouched for existing violations so as to not delete/update due days.

Adds the following for each todo item:

```ts
interface TodoData {
  ...
  daysToWarn?: int;
  daysToError?: int;
}
```

Note: `daysToWarn` and `daysToError` should not be considered in the hash generation for the violation.

**Linter expected implementation should be:**

- If `--update-todo` is passed to the Linter with `daysToWarn` and `daysToError`, `writeTodo()` is called and both arguments are passed. Todos will now have a due date associated to them.
- If error is present in the todo dir, check if `daysToWarn` threshold is reached, report as `warn`, if `daysToError` threshold is reached report as `error`, otherwise report as todo.

Reply:
This seems good. 👍